js_wasm32 cannot run hellope.odin on higher optimization modes

## Context

```
Odin: dev-2022-06-nightly:ba5f7c4e
OS:   Arch Linux, Linux 5.18.3-arch1-1
CPU:  11th Gen Intel(R) Core(TM) i5-11400F @ 2.60GHz
RAM:  15868 MiB
```

## Expected Behavior

"Hellope!" to be displayed in the browser console

## Current Behavior

I get the error `Uncaught (in promise) LinkError: import object field '__truncdfhf2' is not a Function`

## Failure Information (for bugs)

Looking inside the generated .wasm file, it seems it is expecting four functions to be defined in the `env` namespace:

```wasm
...
  (import "env" "__truncdfhf2" (func $__truncdfhf2 (type $t5)))
  (import "env" "__extendhfsf2" (func $__extendhfsf2 (type $t6)))
  (import "env" "__umodti3" (func $__umodti3 (type $t7)))
  (import "env" "__udivti3" (func $__udivti3 (type $t7)))
...
```

This happens with `-opt:1`, `-opt:2`, `-opt:3`, `-o:size`, or `-o:speed`.

When looking at the output without any of these build settings, it seems these functions are simply defined internally:

```wasm
...
  (func $__truncdfhf2 (type $t21) (param $p0 f64) (result i32)
    (local $l1 f32) (local $l2 i32)
    local.get $p0
    f32.demote_f64
    local.set $l1
    local.get $l1
    call $__truncsfhf2
    local.set $l2
    local.get $l2
    return)
...
```

### Steps to Reproduce

1. Simple hellope.odin file:

```odin
package main

import "core:fmt"

main :: proc() {
    fmt.println("Hellope!")
}
```

2. Building it with `odin build . -target:js_wasm32 -o:size` gives a `.wasm` file

3. Try to run the .wasm file using the following html (runtime.js grabbed from vendor:wasm)

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>odin wasm test</title>
</head>
<body>
    <script type="text/javascript" src="runtime.js"></script>
    <script type="text/javascript">
        odin.runWasm(".wasm");
    </script>
</body>
</html>
```

### Failure Logs

![image](https://user-images.githubusercontent.com/7745457/173836969-ff6c4e8f-8fd4-47e3-b1b6-e5e9b234eaa2.png)

Theory likely true: compiling on linux with `odin build . -no-crt -o:speed` gives a linking error:
```
odin_package:(.text+0x2466f): undefined reference to `__truncdfhf2'
```
Of course, this will never be able to run either way because of `-no-crt`, but it *does* sort of compile without optimization.

So in summary, llvm generates calls to specific functions when encountering some operators with 128-bit operands. These are usually defined in compiler-rt / libgcc, but not every architecture has those libraries. I assume this is why odin also defines these functions itself in [core:runtime/internal.odin](https://github.com/odin-lang/Odin/blob/master/core/runtime/internal.odin#L659).

However, right now this only seems to work at the lowest settings of optimization where LLVM does not remove unused definitions. In other cases, these internally-defined procs get removed before they are linked. This is usually unnoticed since they then get linked to compiler-rt / libgcc.

The error happens when odin builds with high optimization and *without* being linked to the C Run Time, which happens when building to WebAssembly or enabling the `-no-crt` flag.

This could potentially be fixed by adding the procs inside internal.odin to the [‘llvm.compiler.used’ Global Variable](https://llvm.org/docs/LangRef.html#id1747), to prevent the compiler from touching them until linking. Given the amount of time I've put into this now today, I might even make my first PR.

--- a/src/llvm_backend_opt.cpp
+++ b/src/llvm_backend_opt.cpp
@@ -380,6 +380,43 @@ void llvm_delete_function(LLVMValueRef func) {
 	LLVMDeleteFunction(func);
 }
 
+void lb_append_to_compiler_used(lbModule *m, LLVMValueRef func) {
+	LLVMValueRef global = LLVMGetNamedGlobal(m->mod, "llvm.compiler.used");
+
+	LLVMValueRef *constants;
+	int operands = 1;
+
+	if (global != NULL) {
+		GB_ASSERT(LLVMIsAGlobalVariable(global));
+		LLVMValueRef initializer = LLVMGetInitializer(global);
+
+		GB_ASSERT(LLVMIsAConstantArray(initializer));
+		operands = LLVMGetNumOperands(initializer) + 1;
+		constants = gb_alloc_array(temporary_allocator(), LLVMValueRef, operands);
+
+		for (int i = 0; i < operands - 1; i++) {
+			LLVMValueRef operand = LLVMGetOperand(initializer, i);
+			GB_ASSERT(LLVMIsAConstant(operand));
+			constants[i] = operand;
+		}
+
+		LLVMDeleteGlobal(global);
+	} else {
+		constants = gb_alloc_array(temporary_allocator(), LLVMValueRef, 1);
+	}
+
+	LLVMTypeRef Int8PtrTy = LLVMPointerType(LLVMInt8TypeInContext(m->ctx), 0);
+	LLVMTypeRef ATy = LLVMArrayType(Int8PtrTy, operands);
+
+	constants[operands - 1] = LLVMConstBitCast(func, Int8PtrTy);
+	LLVMValueRef initializer = LLVMConstArray(Int8PtrTy, constants, operands);
+
+	global = LLVMAddGlobal(m->mod, ATy, "llvm.compiler.used");
+	LLVMSetLinkage(global, LLVMAppendingLinkage);
+	LLVMSetSection(global, "llvm.metadata");
+	LLVMSetInitializer(global, initializer);
+}
+
 void lb_run_remove_unused_function_pass(lbModule *m) {
 	isize removal_count = 0;
 	isize pass_count = 0;
@@ -415,6 +452,7 @@ void lb_run_remove_unused_function_pass(lbModule *m) {
 				Entity *e = *found;
 				bool is_required = (e->flags & EntityFlag_Require) == EntityFlag_Require;
 				if (is_required) {
+					lb_append_to_compiler_used(m, curr_func);
 					continue;
 				}
 			}

